spec/grouping-nodes.md

```diff
@@ -0,0 +1,146 @@
+# Grouping nodes
+
+## Problem description
+
+Today a DataPlaneRole CR defines a grouping of OpenStackDataPlaneNode CRs to
+make it possible to define configuration once but apply it to each Node having
+the same Role. Today a DataPlaneNode can only belong to a single Role. Today
+there is CRD field level inheritance defined between Role and Node. If a CRD
+field is defined both in the Role and the Node then the value from the Node
+will take precedence. With today's model the DataPlaneRole has a direct
+grouping effect for the NovaExternalCompute CRs as well.
+
+However if we move the EDPM related OpenStack service creation from the
+dataplane-operator to the respective service-operator [as proposed in a
+separate spec](https://github.com/openstack-k8s-operators/docs/pull/41), then
+DataPlaneRole based grouping will not be automatically applied. So we need a
+different solution.
+
+Moreover the current restriction that a DataPlaneNode can only belong to a
+single DataPlaneRole is limiting. From nova perspective a set of compute nodes
+can be configured with CPU pinning while from neutron perspective a set of
+compute nodes can be configured with DPDK. If the two sets of nodes overlap but
+are not equal then there will be a need for additional roles to be defined for
+each combination of the configuration (pinning with DPDK, pinning without
+DPDK, sharing with DPDK, sharing without DPDK). This shows that the current
+restriction can lead to the explosion of the number of necessary groups to
+model the different combinations of service configs.
+
+## Proposed change
+
+Allow an EDPM node to belong to multiple groups by defining OpenStack service
+specific profiles. The DataPlaneRole could remain as is today but it would
+represent the grouping of the EDPM nodes only just from the generic
+hardware, host OS, infrastructure configuration perspective.
+
+Each service operator that has EDPM specific service CRD (e.g.
+NovaExternalCompute, OVNMetadataAgent) can define a Profile CRD (e.g.
+ComputeProfile, NetworkProfile) to describe the service specific configuration
+of a set of EDPM nodes. Then the human deployer can use annotations on the
+DataPlaneRole or DataPlaneNode to select a single Profile per service type.
+
+### Example
+
+ComputeProfile with shared CPUs
+```yaml
+apiVersion: nova.openstack.org/v1beta1
+kind: ComputeProfile
+metadata:
+    name: shared-cpus
```

Comment:
To be clear, it is likely this will not be as clear-cut for a real deployment, IIUC. Since you're embedding not just shared/dedicated cpus but *specific* ones, this would end up being something like `dell-r740-shared-cpus` or `hp-dl360g11-shared-cpus`, as different machines from different generations and configurations would end up with different config.

That doesn't change anything about what you're saying here, but I think it's important to note that we'd end up with a lot of ComputeProfile CRs. That's actually a *good* thing IMHO, because it helps to explain the scope of the explosion you mention above. It's not just that currently we'd need to have shared and dedicated versions of each neutron profile, but `hp-dl360g11-shared-dpdk` and `hp-dl360g11-dedicated-dpdk` and `hp-dl360g11-shared-nodpdk` and `hp-dl360g11-dedicated-nodpdk` ... *and* dell variants of each as well :)

Reply:
right since we are not propsoing any composition of multiple profiles e.g. by making the annotation that resfences the profile a list
you would have to create a sperat profile for each hardwre/software combination


with that said we could evolve to enable composition if that was useful but I think there is merit in keeping this simple and just having an option 1:1 mapping rather then starting with many-to-one mapping and all of the precedence issues that would introduce.